api/src/main/java/com/lightbend/lagom/javadsl/api/Descriptor.java

```diff
@@ -390,61 +390,162 @@ public String toString() {
+        TopicCall(TopicId topicId, TopicHolder topicHolder, MessageSerializer<Message, ByteString> messageSerializer, Properties<Message> properties) {
+            this.topicId = topicId;
+            this.topicHolder = topicHolder;
+            this.messageSerializer = messageSerializer;
+            this.properties = properties;
+        }
+
+        /**
+         * Get the topic id.
+         *
+         * @return The topic id
+         */
+        public TopicId topicId() {
+            return topicId;
+        }
+
+        /**
+         * Get the topic holder.
+         *
+         * @return The topic holder.
+         */
+        public TopicHolder topicHolder() {
+            return topicHolder;
+        }
+
+        /**
+         * Get the topic's message serializer.
+         *
+         * @return The message serializer.
+         */
+        public MessageSerializer<Message, ByteString> messageSerializer() {
+            return messageSerializer;
+        }
+
+        /**
+         * Get the properties associated with this topic.
+         *
+         * @return The properties.
+         */
+        public Properties<Message> properties() {
+            return properties;
+        }
+
+        /**
+         * Return a copy of this topic call with the given topic holder configured.
+         *
+         * @param topicHolder The topic holder.
+         * @return A copy of this topic call.
+         */
+        public TopicCall<Message> withTopicHolder(TopicHolder topicHolder) {
+            return new TopicCall<>(topicId, topicHolder, messageSerializer, properties);
+        }
+
+        /**
+         * Return a copy of this topic call with the given message serializer configured.
+         *
+         * @param messageSerializer A
+         * @return A copy of this topic call.
+         */
+        public TopicCall<Message> withMessageSerializer(MessageSerializer<Message, ByteString> messageSerializer) {
+            return new TopicCall<>(topicId, topicHolder, messageSerializer, properties);
+        }
+
+        /**
+         * Return a copy of this copy call with the given property configured.
+         *
+         * @param property The property key.
+         * @param value The value for the property.
+         * @return A copy of this topic call.
+         */
+        public <T> TopicCall<Message> withProperty(Properties.Property<Message, T> property, T value) {
+            return new TopicCall<>(topicId, topicHolder, messageSerializer, properties.withProperty(property, value));
+        }
     }
 
     /**
      * Holds a map of properties.
      */
-    static final public class Properties {
-      public static final class Property<T> {}
-
-      private final PMap<Property<Object>, Object> properties;
-
-      public Properties(PMap<Property<Object>, Object> properties) {
+    public static final class Properties<Message> {
+        /**
+        * A property.
+        */
+        public static final class Property<Message, T> {
```

Comment:
It's so that the partition key strategy can be strongly typed. The partition key strategy is a function of `Message => String`. If property didn't have a type param, then the strategy would either have to be a function of `Object => String`, which would mean implementing it would require the user to do casting, or it would be `Message => String` but the user would have to explicitly set the type, and there would be no way to enforce that the strategy was of the right type when passed to `withProperty`.

The downside of typing it like this is that many properties won't need a type, so it's not needed, but because it is typed and Java doesn't have bottom types (like Scala's Nothing), you can't use a constant for the keys, they always have to be a method that takes the type parameter (like `Collection.empty()`). My thought is that that's a small price to pay for the convenience of the partition key strategy parameter.